Source/WebCore/loader/ResourceMonitorThrottler.cpp

```diff
@@ -81,31 +181,60 @@ void ResourceMonitorThrottler::removeOldestThrottler()
     m_throttlersByHost.remove(oldestKey);
 }
 
-bool ResourceMonitorThrottler::tryAccess(const String& host, ApproximateTime time)
+bool ResourceMonitorThrottler::ThrottlerContainer::tryAccess(const String& host, ContinuousApproximateTime time)
 {
-    if (host.isEmpty())
-        return false;
+    ASSERT(!isMainThread());
+    ASSERT(!host.isEmpty());
 
     auto& throttler = throttlerForHost(host);
-    auto result = throttler.tryAccessAndUpdateHistory(time, m_config);
+    bool wasGranted = throttler.tryAccessAndUpdateHistory(time, m_config);
 
-    if (m_throttlersByHost.size() > m_config.maxHosts) {
-        // Update and remove all expired access times. If no entry in throttler, remove it.
-        m_throttlersByHost.removeIf([protectedThis = Ref { *this }, &time](auto& it) -> bool {
-            return it.value.tryExpire(time, protectedThis->m_config);
-        });
-
-        // If there are still too many hosts, then remove oldest one.
-        while (m_throttlersByHost.size() > m_config.maxHosts)
-            removeOldestThrottler();
+    if (wasGranted) {
+        maintainHosts(time);
+        if (m_persistence)
+            m_persistence->recordAccess(host, time);
     }
+
+    return wasGranted;
+}
+
+void ResourceMonitorThrottler::ThrottlerContainer::clearAllData()
+{
+    if (m_persistence)
+        m_persistence->deleteAllRecords();
+}
+
+void ResourceMonitorThrottler::ThrottlerContainer::maintainHosts(ContinuousApproximateTime time)
```

Comment:
All this `ThrottlerContainer` class should be renamed and moved to its own file. We should never have functions called on various threads in the same file / class.